engine/resources/hetzner.go

```diff
@@ -0,0 +1,710 @@
+	}
+	datacenter, _, err := obj.client.Datacenter.GetByName(ctx, obj.Datacenter)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect Datacenter struct")
+	}
+	// TODO: add more flexible key selection
+	keylist, err := obj.client.SSHKey.All(ctx)
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to collect SSHKey array")
+	}
+	// build serverconfig from given specs & defaults
+	obj.serverconfig = hcloud.ServerCreateOpts{
+		Name:             obj.Name(),        // string
+		ServerType:       serverType,        // *ServerType
+		Image:            image,             // *Image
+		SSHKeys:          keylist,           // []*SSHKey
+		Location:         nil,               // *Location
+		Datacenter:       datacenter,        // *Datacenter
+		UserData:         "",                // string
+		StartAfterCreate: &startAfterCreate, // *bool
+		Labels:           nil,               // map[string]string
+		Automount:        &automount,        // *bool
+		Volumes:          nil,               // []*Volume
+		Networks:         nil,               // []*Network
+		Firewalls:        nil,               // []*ServerCreateFirewall
+		PlacementGroup:   nil,               // *PlacementGroup
+	}
+	// validate resulting config with hloud package
+	if err := hcloud.ServerCreateOpts.Validate(obj.serverconfig); err != nil {
+		return errwrap.Wrapf(err, "invalid serverconfig")
+	}
+	return nil
+}
+
+// getRescueKeys builds a list of keys to be enabled for rescue mode over ssh.
+// HetznerRes.ServerRescueSSHKeys provides the selected keys as []string by name.
+// The corresponding data is collected with the hetzner client (if valid).
+// The resulting keylist is formatted as []*hcloud.SSHKey for use with hcloud,
+// and saved for later use in private field HetznerRes.rescuekeys.
+// TODO: throw error instead of ignoring unknown keynames?
+func (obj *HetznerRes) getRescueKeys(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getRescueKeys()")
+	}
+	var keylist []*hcloud.SSHKey
+	for _, keyname := range obj.ServerRescueSSHKeys {
+		key, _, err := obj.client.SSHKey.GetByName(ctx, keyname)
+		if err != nil {
+			return errwrap.Wrapf(err, "SSHKey GetByName(%s) failed", keyname)
+		}
+		if key != nil {
+			keylist = append(keylist, key)
+			if obj.init.Debug {
+				obj.init.Logf("append known keyname: %s", key)
+			}
+		}
+		if obj.init.Debug {
+			obj.init.Logf("ignore unknown keyname: %s", key)
+		}
+	}
+	obj.rescuekeys = keylist
+	return nil
+}
+
+// waitForSteadyState polls for serverupdates from the hetzner api until
+// context cancellation, an error occurs, or when the target server reaches
+// a steady state ("running" or "off").
+func (obj *HetznerRes) waitForSteadyState(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("waitForSteadyState()")
+	}
+	for {
+		if obj.server == nil {
+			return fmt.Errorf("server is unavailable")
+		}
+		if obj.server.Status == hcloud.ServerStatusRunning {
```

Comment:
is it not possible for obj.server.Status to be in a third or fourth state? If so, do we want to know this? Perhaps it's something like "BuildingOnFire" ... so we'd want to not wait forever in that state... But we don't know it doesn't exist as a state yet? So perhaps we want a switch here, and a default: of "unknown state" and an error??

Reply:
Hmm, this made me realize that I'm not explicitly looking at any of the transient states, even though some of them could be early indicators of trouble... Converting the current functionality here to a switch seems mostly a matter of style preference, but I agree that it could provide a cleaner framework to catch exceptions if any bugs ever need to be addressed. 

Will make the simple transition to a switch case for now, but I need to think a bit longer about those transient states.

Reply:
Added error cases for the `unknown` and `deleting` states, since those really shouldn't happen there. 
Beyond that there shouldn't be many issues, it just needs to wait until the transients converge into a valid steady state.